src/passes/CodePushing.cpp

```diff
@@ -237,6 +244,178 @@ class Pusher {
+        // move past it. Note the effects and continue.
+        cumulativeEffects.walk(list[i]);
+        continue;
+      }
+
+      auto index = pushable->index;
+
+      auto iter = pushableEffects.find(pushable);
+      if (iter == pushableEffects.end()) {
+        iter = pushableEffects
+                 .emplace(std::piecewise_construct,
+                          std::forward_as_tuple(pushable),
+                          std::forward_as_tuple(passOptions, module, pushable))
+                 .first;
+      }
+      auto& effects = iter->second;
+
+      if (cumulativeEffects.invalidates(effects)) {
+        // This can't be moved forward. Add it to the things that are not
+        // moving.
+        cumulativeEffects.walk(list[i]);
+        continue;
+      }
+
+      // We only try to push into an arm if the local is used there. If the
+      // local is not used in either arm then we'll want to push it past the
+      // entire if, which is what optimizeSegment handles.
+      //
+      // We can push into the if-true arm if the local cannot be used if we go
+      // through the other arm:
+      //
+      //    x = op();
+      //    if (..) {
+      //      // we would like to move "x = op()" to here
+      //      ..
+      //    } else {
+      //      use(x);
+      //    }
+      //    use(x);
+      //
+      // Either of those use(x)s would stop us from moving to if-true arm.
+      //
+      // One specific case we handle is if there is a use after the if but the
+      // arm we don't push into is unreachable. In that case we only get to the
+      // later code after going through the reachable arm, which is ok to push
+      // into:
+      //
+      //    x = op();
+      //    if (..) {
+      //      // We'll push "x = op()" to here.
+      //      use(x);
+      //    } else {
+      //      return;
+      //    }
+      //    use(x);
+      auto maybePushInto = [&](Expression*& arm,
+                               const Expression* otherArm,
+                               const EffectAnalyzer& armEffects,
+                               const EffectAnalyzer& otherArmEffects) {
+        if (!arm || !armEffects.localsRead.count(index) ||
+            otherArmEffects.localsRead.count(index)) {
+          // No arm, or this arm has no read of the index, or the other arm
+          // reads the index.
+          return false;
+        }
+        if (postIfEffects.localsRead.count(index) &&
+            (!otherArm || otherArm->type != Type::unreachable)) {
+          // The local is read later, which is bad, and there is no unreachable
+          // in the other arm which as mentioned above is the only thing that
+          // could have made it work out for us.
+          return false;
+        }
+
+        // We can do it! Push into one of the if's arms, and put a nop where it
+        // used to be.
+        Builder builder(module);
+        auto* block = builder.blockify(arm);
+        arm = block;
```

Comment:
It doesn't look like the separate `block` variable adds much here.
```suggestion
        arm = builder.blockify(arm);
```

Reply:
We need the block on the next line,
```cpp
        ExpressionManipulator::spliceIntoBlock(block, 0, pushable);
```
That function takes a `Block` as its input.